lib/strings.nix

```diff
@@ -1080,6 +1080,67 @@ rec {
   */
   escapeShellArgs = concatMapStringsSep " " escapeShellArg;
 
+  /**
+    This is like `escapeShellArg` except it won't quote the argument if not
+    necessary.
+
+    # Inputs
+
+    `string`
+    : 1\. Function argument
+
+    # Type
+
+    ```
+    escapeBashArg :: string -> string
+    ```
+
+    # Examples
+    :::{.example}
+    ## `lib.strings.escapeBashArg` usage example
+
+    ```nix
+    escapeBashArg "foo'bar"
+    => "'foo'\\''bar'"
+    escapeBashArg "foobar"
+    => "foobar"
+    ```
+
+    :::
+  */
+  escapeBashArg = arg:
+    if match ".*[]\\[\"'[:space:]\\\\#$&(){}<>|;*?=!~].*" arg == null
```

Comment:
I'm having a hard time grasping this regex. What's going on with `[]\\[...]`?

And it would be great to link to some reference showing that these are all the characters that need to be escaped, so that we can be sure we aren't missing any. In particular I'm also wondering about things like newlines, random ASCII, but also Unicode. And if we can't make a reasonable and reliable deny-list of characters, an allow-list might be a better choice instead.

Reply:
I agree that I’d feel more comfortable with something like `[a-zA-Z0-9,._+:@%/-]` (from https://stackoverflow.com/questions/15783701/which-characters-need-to-be-escaped-when-using-bash) to handle the majority of cases here without any risk. Although it’s a little sad that any non‐ASCII letters will end up escaped that way.

Reply:
How about `[[:alnum:],._+:@%/-]`?

Reply:
Sounds good to me!

Reply:
Will that work? Nix strings don't natively support Unicode, and regex handling is dedicated to the native C++ library with platform-specific behaviour(!), so I wonder how it will behave. In principle it sounds good to me, though.

Reply:
Part of the reason I suggested using `[:alnum:]` is because I think that will be more robust in the presence of locales.  Keep in mind that in a bracket expression `a-z` is also locale-sensitive and provides fewer guarantees about what it actually matches:

> Within a bracket expression, a range expression consists of two characters separated by a hyphen. It matches any single character that sorts between the two characters, inclusive. In the default C locale, the sorting sequence is the native character order; for example, ‘[a-d]’ is equivalent to ‘[abcd]’. In other locales, the sorting sequence is not specified, and ‘[a-d]’ might be equivalent to ‘[abcd]’ or to ‘[aBbCcDd]’, or it might fail to match any character, or the set of characters that it matches might be erratic, or it might be invalid

Reply:
Oh, joy… 🫠

I believe Nix uses `std::regex`; no idea how that interacts with locales, but `[:alnum:]` seems reasonable to me. If you could add some tests for Unicode edge‐cases that’d be great assurance.